server/src/main/java/org/opensearch/tasks/Task.java

```diff
@@ -195,6 +232,101 @@ public Status getStatus() {
         return null;
     }
 
+    /**
+     * Returns thread level resource consumption of the task
+     */
+    public Map<Long, List<ThreadResourceInfo>> getResourceStats() {
+        return resourceStats;
+    }
+
+    /**
+     * Returns total resource usage of the task
+     */
+    public TaskResourceUsage getTotalResourceStats() {
+        return new TaskResourceUsage(getTotalResourceUtilization(ResourceStats.CPU), getTotalResourceUtilization(ResourceStats.MEMORY));
+    }
+
+    /**
+     * Returns total resource consumption for a specific task stat.
+     */
+    public long getTotalResourceUtilization(ResourceStats taskStats) {
+        long totalResourceConsumption = 0L;
+        for (List<ThreadResourceInfo> threadResourceInfosList : resourceStats.values()) {
+            for (ThreadResourceInfo threadResourceInfo : threadResourceInfosList) {
+                for (Map.Entry<ResourceStatsType, ResourceUsageInfo> entry : threadResourceInfo.getResourceUsageInfos().entrySet()) {
+                    if (entry.getKey().isOnlyForAnalysis() == false) {
+                        totalResourceConsumption += entry.getValue().getStatsInfo().get(taskStats).getTotalValue();
+                    }
+                }
+            }
+        }
+        return totalResourceConsumption;
+    }
+
+    /**
+     * Add thread's resource consumption information
+     * @param threadId ID of the thread
+     * @param statsType stats type
+     * @param resourceUsageMetrics resource consumption metrics of the thread
+     */
+    public void addTaskResourceStats(long threadId, ResourceStatsType statsType, ResourceUsageMetric... resourceUsageMetrics) {
+        if (statsType != ResourceStatsType.WORKER_STATS) {
+            throw new IllegalArgumentException("Adding thread resource information should always have WORKER_STATS as stats type");
+        }
+        final List<ThreadResourceInfo> threadResourceInfoList = resourceStats.computeIfAbsent(threadId, k -> new ArrayList<>());
```

Comment:
Lets move this method out to a separate service, something like `TaskReourceTrackingService`

Reply:
@Bukhtawar Why we need a separate service to keep track of all the tasks related resources ? TaskManager currently already keeps track of all the tasks and it is the component which manages the lifecycle of the tasks as well. With Tasks exposing the APIs to update/add the resource stats seems like a good fit to me. The caller can use TaskManager to get the corresponding task object and update the resource stats for that task using Task object API.

Reply:
Here is my understanding. `Task` is just a container that holds attributes that defines a task(task information). It has no business logic whatsoever. `TaskManager` orchestrates the lifecycle of the task, however that doesn't mean we overload functionalities in the same class. Non-essential add-ons that are not critical in orchestrating a task makes more sense to be separate decoupled and abstracted away like `TaskResultsService` which tries to store the result of the task, but doesn't directly influence the task orchestration.  

There would be more add-ons I envision like `SlowTaskTrackingService` that monitors all slow task execution and takes action like logging/metrics etc. 

With that in mind I think we the current implementation overloads `Task` or for that matter `TaskManager`

Reply:
`ResourceStats` which is being added is an attribute of the task and that's why it is put within `Task` object. If you see other `Task` types like `CancellableTask` they do have business logic specific to the attribute it maintains. Here, `ResourceStats` is a common attributes which will be used by all the tasks type hence in base class. `TaskManager` doesn't only orchestrates the lifecycle but it also owns other logic like to `storeResult` or `cancelTask` which the individual service uses. Like `TaskResultsService` is called from within the `TaskManager` to store the result since the logic to persist the result is abstracted in that service, however for each task it is again called via `TaskManager` and the logic to get the results to be stored is in `Task` object itself. Similarly for capturing resource usage we will have 2 mechanisms: 1) The worker which has the task information which it is working on and will update the `Task` object directly after fetching it via `TaskManager`. 2) Periodic refresh which can again be within `TaskManager` as I don't see any heavy business logic for that purpose compared to `TaskResultsService` or `TaskCancellationService`. If we plan to add any heavier logic then it can follow something like below observer model.

Yes add-ons like `SlowTaskTrackingService` can be abstracted out which can keep track of all or subset of tasks during register/unregister call from `TaskManager`. We plan to add these in follow-up PR. Those service will consume the `Task` object and based on the `resourceStats` or other attributes can decide to either update the sink (logger / system index / etc) or skip it. The business logic whether to add information for such tasks in sink or not will be abstracted in those services similar to `TaskResultsService`

Reply:
There are two thing, the `ResourceStats ` which can definitely be a part of `Task` but the how part of getting the stats should be abstracted outside it. `TaskManager` can help coordinate/manage the same, but I would like update and store logic to be sitting elsewhere so as to keep `TaskManager` light-weight

Reply:
`ResourceStats` being an attribute of `Task` there is a need to provide `get/update` logic in `Task` class itself to manipulate or retrieve that attribute. Now coming to `TaskManager` providing the management of these `Task` attribute is preferred, since then all the updates to Task attributes will be in a single place. If you have a separate service like `TaskResourceService` then that will expose the same methods as `TaskManager` and will internally call the `TaskManager` methods. In both the models, we cannot escape the methods in `TaskManager` without keeping a copy of `Task` objects reference which I want to avoid.

As far as I/O stats are concerned that is another type of resource stats and should be extendable like CPU and Memory. It will not impact the update methods to add information about new stats. But for now let us not think about what future enhancements can come in, I am sure it can be easily extended.

@andrross and @dblock : Since you guys have reviewed the PR too, would like to hear your thoughts as well.